RemoteLayerBackingStore copies unnecessary pixels from back buffer to front buffer.

mattwoodrow opened the pull request:
#### c26d9ad7d08d34a9c2d515fbe6781ae94bf3565f
<pre>
RemoteLayerBackingStore copies unnecessary pixels from back buffer to front buffer.
<a href="https://bugs.webkit.org/show_bug.cgi?id=256621">https://bugs.webkit.org/show_bug.cgi?id=256621</a>
&lt;rdar://109179068&gt;

Reviewed by Simon Fraser.

If the new front buffer is the old back buffer, then we only need to copy the area of pixels that we painted
last time in order to make it match the new back buffer. Take area, subtract the pixels we&apos;re going to overwrite
for this update, and only copy that area.
Uses IntRects as a pessimistic superset of the required area, rather than Regions, to ensure we don&apos;t spend extra time
clipping (which could potentially take longer than the saved time from smaller copies).

Also detects when we have a newly allocated buffer, which are cleared during allocation, and avoids clearing them a
second time.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::applySwappedBuffers):
(WebKit::RemoteLayerBackingStore::ensureFrontBuffer):
(WebKit::RemoteLayerBackingStore::drawInContext):

Canonical link: <a href="https://commits.webkit.org/263994@main">https://commits.webkit.org/263994@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ce73f7a5ca5e4269f64266ef29051cbf65a53013

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6310 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6506 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6685 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7884 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6636 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6726 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6455 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9527 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6422 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6428 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5715 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7954 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3900 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13577 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5764 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5774 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/8021 "10 api tests failed or timed out") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6259 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5118 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5670 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9825 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/751 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6040 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
```diff
@@ -560,8 +568,17 @@ static bool hasValue(const ImageBufferBackendHandle& backendHandle)
+            context.drawImageBuffer(*m_backBuffer.imageBuffer, { 0, 0 }, { CompositeOperator::Copy });
+        else {
+            Region copyRegion(*previouslyPaintedRect);
+            copyRegion.subtract(m_dirtyRegion);
```
I wonder if it’s worth doing the Region math here. Maybe just a rect intersection of `dirtyBounds` and `* previouslyPaintedRect` would be enough?

mattwoodrow replied:
Intersection would give the wrong rectangle wouldn't it? We want to copy the pixels that aren't in dirtyBounds, since those will be repainted this frame.

smfr replied:
Yeah that wasn't right. There's probably an approximation you could do with rects, but there are only a few rects here so the region math should be cheap enough.

smfr commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
```diff
@@ -207,6 +208,8 @@ class RemoteLayerBackingStore {
     Buffer m_backBuffer;
     Buffer m_secondaryBackBuffer;
 
+    std::optional<WebCore::IntRect> previouslyPaintedRect;
```
`m_ previouslyPaintedRect`

mattwoodrow committed:
RemoteLayerBackingStore copies unnecessary pixels from back buffer to front buffer.
https://bugs.webkit.org/show_bug.cgi?id=256621
<rdar://109179068>

Reviewed by Simon Fraser.

If the new front buffer is the old back buffer, then we only need to copy the area of pixels that we painted
last time in order to make it match the new back buffer. Take area, subtract the pixels we're going to overwrite
for this update, and only copy that area.
Uses IntRects as a pessimistic superset of the required area, rather than Regions, to ensure we don't spend extra time
clipping (which could potentially take longer than the saved time from smaller copies).

Also detects when we have a newly allocated buffer, which are cleared during allocation, and avoids clearing them a
second time.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::applySwappedBuffers):
(WebKit::RemoteLayerBackingStore::ensureFrontBuffer):
(WebKit::RemoteLayerBackingStore::drawInContext):

Canonical link: https://commits.webkit.org/263994@main
